src/process_user_code.py

```diff
@@ -43,16 +46,39 @@ def run(self):
             read_val = sys.stdin.readline()
             sys.stdin.flush()
             try:
-                new_state = json.loads(read_val)
-                for event in CONSTANTS.EXPECTED_INPUT_EVENTS_CPX:
-                    cpx._Express__state[event] = new_state.get(
-                        event, cpx._Express__state[event]
-                    )
+                new_state_message = json.loads(read_val)
+                device = new_state_message.get(CONSTANTS.ACTIVE_DEVICE_FIELD)
+                new_state = new_state_message.get(CONSTANTS.STATE_FIELD, {})
+
+                if device == CPX:
+                    update_cpx(new_state)
+                elif device == MICROBIT:
+                    update_microbit(new_state)
+                else:
+                    raise Exception(CONSTANTS.DEVICE_NOT_IMPLEMENTED_ERROR)
 
             except Exception as e:
                 print(CONSTANTS.ERROR_SENDING_EVENT, e, file=sys.stderr, flush=True)
 
 
+def update_cpx(new_state):
+    for event in CONSTANTS.EXPECTED_INPUT_EVENTS_CPX:
+        cpx._Express__state[event] = new_state.get(event, cpx._Express__state[event])
+
+
+def update_microbit(new_state):
```

Comment:
I think it would make more sense to handle the state changing logic inside each component, instead of having all the different components update state in this method. What do you think?

Reply:
that's true... I'll try implementing that for the next commit!

Reply:
implemented!